src/server.js

```diff
@@ -23,39 +26,78 @@ class ExecSession {
       Cmd: options.command,
     };
     this.attachOptions = {
+      stream: true,
       stdin: true,
       stdout: true,
       stderr: true,
-      stream: true,
     };
     this.container = options.container;
-    this.socket = options.socket;
     this.server = options.server;
+    this.socket = options.socket;
+    this.closed = false;
+
+    //must set this before this.execute() in case of premature close
+    this.socket.on('close', () => {
+      //should be how to ctrl+c ctrl+d, might be better way to kill
+      debug('client close');
+      //for now, it kills this session
+      this.close();
+    });
   }
 
-  async execute () {
+  async execute() {
     //TODO: add error handling support here
     this.exec = await this.container.exec(this.execOptions);
     this.execStream = await this.exec.start(this.attachOptions);
+    // this.execStream = await new Promise(async (accept, reject) => {
+    //   var req = http.request({
+    //     socketPath: '/var/run/docker.sock', //replace with generic later
+    //     path: '/exec/' + this.exec.id + '/start?' + qs.stringify(this.attachOptions),
+    //     method: 'POST',
+    //     headers: {
+    //       'Content-Type': 'application/json',
+    //       'Connection': 'upgrade',
+    //       'Upgrade': 'tcp',
+    //     }
+    //   }, reject);
+    //   req.write(JSON.stringify({
+    //     Detach: false,
+    //     Tty: this.options.tty,
+    //   }));
+    //   req.end();
+    //   req.on('upgrade', accept);
+    //   req.on('error', reject);
+    //   req.setTimeout(10000, reject);
+    // });
+    // debug(this.execStream);
+    // this.execStream.write(new Buffer([30]));
 
-    //handling output
-    this.strout = through((data) => {
-      this.sendMessage(msgcode.stdout, new Buffer(data));
-    });
+    var header = null;
 
-    this.strerr = through((data) => {
-      this.sendMessage(msgcode.stderr, new Buffer(data));
-    });
+    this.execStream.on('readable', () => {
+      header = header || this.execStream.read(8);
+      while (header !== null) {
+        var type = header.readUInt8(0);
+        var payload = this.execStream.read(header.readUInt32BE(4));
+        if (payload === null) {
+          break;
+        }
+        if (!this.sendMessage(type, payload)) {
+          this.execStream.pause();
+          this.strbuf.on('drain', this.execStream.resume);
+        }
 
-    this.exec.modem.demuxStream(this.execStream, this.strout, this.strerr);
+        //try to set new header to continue reading
+        header = this.execStream.read(8);
+      }
+    });
     //This stream is created solely for the purposes of pausing, because
     //data will only buffer up in streams using this.queue()
-    this.strbuf = through();
+    this.strbuf = through2();
 
-    const MAX_OUTSTANDING_BYTES = 8 * 1024 * 1024;
     this.outstandingBytes = 0;
 
-    this.strbuf.pipe(through((data) => {
+    this.strbuf.pipe(through2((data, enc, cb) => {
```

Comment:
Don't pipe.. this is bad... if you wanted to you should use node flow control... not call pause/resume..
How about
```js
this.strbuf.on('data', (data) => {
  .... // call strbuf.pause/resume as needed
});
this.strbuf.resume(); // or maybe not do that here... but wait for client to send `resume`...
```